### ISSUE TITLE
[decorator] loops in function findWithRegex in tweet example

### ISSUE THREAD

**Issue by VanishingDante** [facebook/draft-js#842]
```
      function findWithRegex(regex, contentBlock, callback) {
        const text = contentBlock.getText();
        let matchArr, start;
        while ((matchArr = regex.exec(text)) !== null) {
          start = matchArr.index;
          callback(start, start + matchArr[0].length);
        }
      }
```

above's the function used in the tweet example, 
`regex` and `text` are both constants, shouldn't it loop forever?

if I add `console.log` to it, the log method will **only exec once**
but If i add `break` statement to the loop, this example will not behave normally

why?

**Comment by arturmuller** [comment]
Hey. Not sure what you mean by behaving normally. As for your other points:

In JS, constants cannot be reassigned, but can be mutated. For example:

```js
// this is totally valid
const myObject = {}
myObject.foo = "bar"

// this is not, and it will throw
const myObject2 = {}
myObject2 = "quux"
```

The `exec` method changes the `lastIndex` property on the regex object. Every time you call it, it will return the next match from the string or `null` if there was nothing found. You can find more info here:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec

Hope that sheds some light on what is happening in those lines of code. Admittedly, it **is** pretty confusing. :)

---

As a side-note I think you'd have better luck getting these kinds of questions answered on StackOverflow as it is more general JS rather then something specific to this library.

**Comment by VanishingDante** [comment]
Thank you @arturmuller !

suppose there will be exactly one matched array in the current contentBlock text, like this `@draft`. The text will never be like `@draft @dra` or sth will match more than once or less than once.

I add `let i = 0` and `console.log(i++) ` to it

```
      function findWithRegex(regex, contentBlock, callback) {
        const text = contentBlock.getText();
        let matchArr, start;
        let i = 0;
        while ((matchArr = regex.exec(text)) !== null) {
          console.log(i++);
          start = matchArr.index;
          callback(start, start + matchArr[0].length);
        }
      }
```

every time I type a letter (or the text in this block changes? contentBlock changes?), this function will be called, and print **one** `0`.

The `console.log` function shows that, the `while` code block only loops once in such situation, so I think the above `findWithRegex` function should equal to below

```
      function findWithRegex(regex, contentBlock, callback) {
        const text = contentBlock.getText();
        let matchArr, start;
        let i = 0;
        while ((matchArr = regex.exec(text)) !== null) {
          console.log(i++);
          start = matchArr.index;
          callback(start, start + matchArr[0].length);
          break;
        }
      }
```

the only difference is the `break`.

to my surprise, with the `break`, 
every time I type a letter (or the text in this block changes? contentBlock changes?), the `findWithRegex` is called,
but every **twice**, the `while` code block will loop once
 [demo's here](https://jsbin.com/qocaruxiji/1/edit?html,console,output)

I'm wondering why

**Comment by davidchang** [comment]
Hi @VanishingDante - admittedly, the example is pretty hard to follow, but this type of question is better geared towards something a more JS oriented forum (perhaps about understanding regex.exec) rather than as a Draft.js issue (as it's not specific to Draft.js). [Though the actionable item here could be "use simpler regular expression parser in examples"]

Decorator code is going to be called regularly upon changes to detect whether or not the UI needs to be updated to match the new ContentState. Does this make sense? If you're breaking within the while, the code within the while loop will only run once.
